Create Tune API in the Katib SDK

andreyvelich opened the pull request:
Fixes: https://github.com/kubeflow/katib/issues/1585.
Inspired by [KFP `create_component_from_func`](https://github.com/kubeflow/pipelines/blob/master/sdk/python/kfp/components/component_factory.py).

This API will allow user to create Katib HP Tuning Experiment without building the Experiment YAML.
This is the first small step to simplify our Kubeflow SDKs and to avoid Kubernetes complexity.

Later we can extend this functionality (support more Katib features in a simple way), give more spec options via APIs.

Also, we can provide API (e.g. `katib.list_base_images`) to show list of supported images (e.g. `BASE_IMAGE_TENSORFLOW`, `BASE_IMAGE_MXNET`) with their description.


cc @kubeflow/wg-training-leads @tenzen-y @anencore94 Please give your feedback on the API design.

andreyvelich committed:
Create Tune API in the Katib SDK

andreyvelich commented:
/hold Hold until https://github.com/kubeflow/training-operator/pull/1659 is merged.

tenzen-y commented:
Thanks for implementing the new feature!
First, I'll review the PR submitted in training-operator repo.

anencore94 commented on README.md:
```diff
@@ -189,10 +189,51 @@ katib-ui-5767cfccdc-pwg2x           1/1     Running     0          36s
+
+# Step 1. Create an objective function.
+def objective(parameters):
+    # Import required packages.
+    import time
```
I understand why this kind of logic is necesary, but we might find out more friendly way to specify packages by users.
However, as a minimal feature, this is ok if we could instruct users that they couldn't make a mistake to miss required packages in the objective function.

andreyvelich replied:
I agree, but I think Kubeflow Pipelines follows the same way: https://www.kubeflow.org/docs/components/pipelines/sdk/python-function-components/#building-python-function-based-components
@zijianjoy Do you know if Pipelines SDK verifies that all function imports are defined properly ?

@anencore94 Any ideas how to verify this?
cc @tenzen-y @gaocegege @johnugeorge

anencore94 replied:
We may need to automatically validating this after this PR. How about making an issue for tracking this feature now ?
If we could make an idea for this validation in general, we may apply this for other kf components like pipelines and training-operators and so on.

andreyvelich replied:
Sure, that sound good @anencore94.

anencore94 commented on sdk/python/v1beta1/kubeflow/katib/api/katib_client.py:
```diff
@@ -430,7 +701,56 @@ def get_optimal_hyperparameters(self, name=None, namespace=None):
         return result
+
+
+def double(min: float, max: float, step: float = None):
```
How about gather these hyperparameter helper function to another inner module?
For example, the users could call these functions by

```python
import kubeflow.katib as katib

hp_search_space = {
    "a": katib.hp_search_space.int(min=10, max=20),
    "b": katib.hp_search_space.double(min=0.1, max=0.2)
}

katib_client.tune(
    name=name,
    objective=objective,
    hp_search_space=hp_search_space,
    objective_metric_name="result",
    max_trial_count=12
)
```

IMO, importing by `katib.int` makes me it's too board that we couldn't see it is for hyperparameter search space type immediately.

andreyvelich replied:
I was even thinking to change this API in the future to support various distributions: `uniform`, `loguniform`, `categorical`, etc.

FYI, hyperopt uses [`hp` module](https://github.com/kubeflow/katib/blob/master/pkg/suggestion/v1beta1/hyperopt/base_service.py#L59), skop uses [`space` module](https://github.com/kubeflow/katib/blob/master/pkg/suggestion/v1beta1/skopt/base_service.py#L54), optuna uses [`distributions` module](https://github.com/kubeflow/katib/blob/master/pkg/suggestion/v1beta1/optuna/base_service.py#L94), Ray uses [tune module](https://docs.ray.io/en/latest/tune/api_docs/suggestion.html?highlight=search#ax-tune-search-ax-axsearch). I think we can learn something from it. 

`search` module looks good to me, but I am not sure what is the best UX here.

What others think @terrytangyuan @gaocegege @johnugeorge @c-bata @a9p @g-votte @tenzen-y ?

terrytangyuan replied:
Ref https://github.com/kubeflow/katib/pull/1951#discussion_r971944632. 

IMO it feels more natural if we could support built-in lists or common numpy arrays, etc. Is there a particular motivation for bringing a separate module in the Katib SDK just to do the same thing?

andreyvelich replied:
@terrytangyuan List and numpy arrays might not be enough. 
How we can identify search space for `uniform` or `loguniform` in the future?

e.g. Hyperopt utilise some numpy APIs, but it still has it's own functions for the search space: https://github.com/hyperopt/hyperopt/blob/master/hyperopt/pyll/stochastic.py#L41-L43

terrytangyuan replied:
The functions in that module seem to be very thin wrappers. If we could completely avoid creating additional wrappers/layers, it would be great.

tenzen-y replied:
I agree with @andreyvelich.
In this PR, we can start in the minimal feature. So, we can support other distributions in other PRs.

andreyvelich replied:
@terrytangyuan Do you have any ideas how we can improve UX on this ?

terrytangyuan replied:
I am ok as is if other projects are following the same convention.

anencore94 commented on sdk/python/v1beta1/kubeflow/katib/constants/constants.py:
```diff
@@ -15,13 +15,21 @@
-EXPERIMENT_PLURAL = 'experiments'
-SUGGESTION_PLURAL = 'suggestions'
-TRIAL_PLURAL = 'trials'
+KUBEFLOW_GROUP = "kubeflow.org"
+EXPERIMENT_KIND = "Experiment"
+EXPERIMENT_PLURAL = "experiments"
+SUGGESTION_PLURAL = "suggestions"
+TRIAL_PLURAL = "trials"
```
how about speficy these variables to type hint `Final[str]`?

anencore94 commented on sdk/python/v1beta1/kubeflow/katib/api/katib_client.py:
```diff
@@ -430,7 +701,56 @@ def get_optimal_hyperparameters(self, name=None, namespace=None):
+    return parameter
+
+
+def categorical(list: List):
```
do we need to check if list is `[]` or not ?

andreyvelich replied:
Should we introduce this check in [Validation Webhook](https://github.com/kubeflow/katib/blob/master/pkg/webhook/v1beta1/experiment/validator/validator.go#L244-L246), so we can have single place for spec validation ?

anencore94 replied:
Of course but I'm not sure if we need some fast-validation for the user who are not familiar with katib. Also representing the validation rule in python's stdout would be better for them to find out a mistake.

andreyvelich replied:
I think if Validation Webhook fails, user also will see an error in the Python's stdout before creating Katib Experiment.
What do you think @anencore94 ?

anencore94 replied:
If the validadation webhook's failure message is detailed enough, your suggestion would be much better than implementing the same validation check in both!

anencore94 commented on sdk/python/v1beta1/kubeflow/katib/api/katib_client.py:
```diff
@@ -53,47 +60,253 @@ def _is_ipython(self):
+            parallel_trial_count: Number of Trials that Experiment runs in parallel.
+            max_failed_trial_count: Maximum number of Trials allowed to fail.
+            retain_trials: Whether Trials' resources (e.g. pods) are deleted after Succeeded state.
+            packages_to_install: List of Python packages to install before
+                executing the objective function.
```
how about saying that these field is for packages except already installed in the `base_image` ?

andreyvelich replied:
Sure. @anencore94 What do you think about this:
```
List of Python packages to install in addition to base image packages. These packages are installed before executing the objective function.
        ```

anencore94 replied:
Looks nice :) Thanks

anencore94 commented on sdk/python/v1beta1/kubeflow/katib/api/katib_client.py:
```diff
@@ -53,47 +60,253 @@ def _is_ipython(self):
+        additional_metric_names: List[str] = [],
+        objective_type: str = "maximize",
+        objective_goal: float = None,
+        max_trial_count: int = None,
+        parallel_trial_count: int = None,
```
how about these fields' default values to `1` for who are not familiar with katib experiments?

andreyvelich replied:
We set default Trial Threshold here: https://github.com/kubeflow/katib/blob/master/pkg/apis/controller/experiments/v1beta1/experiment_defaults.go#L35-L39
I am not sure, if we want to introduce another default value in the `tune` API.
WDYT @anencore94 @tenzen-y ?

anencore94 replied:
I introduce this because of the same reason with this comment: https://github.com/kubeflow/katib/pull/1951#discussion_r972651507.

I'm not sure if we have to implement such fast-validation (which is duplicated) or not.

tenzen-y replied:
Since introducing duplicated values will cause the bug, it would be good not to set the default values in this.

anencore94 replied:
I agree with you, but then I think we should write docstring and tutorial page more detailed and kind!

andreyvelich replied:
@anencore94 @tenzen-y Maybe we update docstring with link to [this page](https://www.kubeflow.org/docs/components/katib/experiment/#configuration-spec) for Trial budget info?

tenzen-y replied:
That sounds good.

anencore94 commented on sdk/python/v1beta1/kubeflow/katib/api/katib_client.py:
```diff
@@ -53,47 +60,253 @@ def _is_ipython(self):
+
+        Args:
+            name: Name for the Experiment.
+            objective: Objective function that Katib uses to train the model.
+                This function must be Callable and it must have only one dict argument.
+                Katib uses this argument to send HyperParameters to the function.
```
I think we need to strongly-specify that objective function should inner-import statements for all necesary packages

andreyvelich replied:
Let's discuss about it in this comment: https://github.com/kubeflow/katib/pull/1951#discussion_r972199141.
At the minimum, we can just extend this comment as you mentioned.

anencore94 commented on sdk/python/v1beta1/kubeflow/katib/api/katib_client.py:
```diff
@@ -53,47 +60,253 @@ def _is_ipython(self):
+        algorithm_name: str = "random",
+        objective_metric_name: str = None,
+        additional_metric_names: List[str] = [],
+        objective_type: str = "maximize",
```
maybe we could use "maximize" and "minimize" as an `Enum` class or `Literal`

andreyvelich replied:
I was also thinking about it. 
Is there a way to automatically generate enums for our [Python Models](https://github.com/kubeflow/katib/blob/master/sdk/python/v1beta1/kubeflow/katib/models/v1beta1_objective_spec.py#L40) ?
In our APIs `type` param is enum: https://github.com/kubeflow/katib/blob/master/pkg/apis/controller/common/v1beta1/common_types.go#L96

anencore94 replied:
By "automatically" means that automatically generate python Enum class from the golang file or openapi.json ?

andreyvelich replied:
> By "automatically" means that automatically generate python Enum class from the golang file or openapi.json ?

Exactly. I am not sure if `kube-openapi` supports enum generation. Check this: https://github.com/kubernetes/enhancements/issues/2887.
Maybe we should update `kube-openapi` version for this feature.

anencore94 replied:
I have no many experience with it sadly. How about make an issue with this feature, and merge this PR with hardcoded for now ?

andreyvelich replied:
Sounds good @anencore94 !

anencore94 commented on sdk/python/v1beta1/kubeflow/katib/api/katib_client.py:
```diff
@@ -53,47 +60,253 @@ def _is_ipython(self):
+        if max_failed_trial_count is not None:
+            experiment.spec.max_failed_trial_count = max_failed_trial_count
+
+        # Modify objective function.
+        # Check if objective function is callable.
+        if not callable(objective):
+            raise ValueError(
+                f"Objective function must be callable, got function type: {type(objective)}"
+            )
+
+        # Verify the objective function arguments.
+        objective_signature = inspect.signature(objective)
+        try:
+            objective_signature.bind({})
+        except Exception:
+            raise ValueError(
+                "Invalid args in the Objective function. "
+                "The function args must have only 'parameters' dictionary. "
+                f"Current Objective arguments: {objective_signature}"
+            ) from None
```
How about extracting this part to a single function with naming something like `validate_objective_fucntion`?
I think this kinds of validation for user's wrong-use could become bigger and bigger.

anencore94 commented on sdk/python/v1beta1/kubeflow/katib/api/katib_client.py:
```diff
@@ -53,47 +60,253 @@ def _is_ipython(self):
+        # def objective(parameters):
+        #     print(f'Parameters are {parameters}')
+        # objective({'lr': '${trialParameters.lr}', 'epochs': '${trialParameters.epochs}', 'is_dist': False})
+        objective_code = f"{objective_code}\n{objective.__name__}({input_params})\n"
```
I'm not sure do we need the last `\n` or not

andreyvelich replied:
I think that is required for the newline in `ephemeral_objective.py` file. It's good practice to have it.

anencore94 reviewed (commented):
Thanks for awesome feature !!! users could generate katib experiment much easier now :)
I left some personal opinions.

terrytangyuan commented on sdk/python/v1beta1/kubeflow/katib/constants/constants.py:
```diff
@@ -15,13 +15,21 @@
+
+# Supported base images for the Katib Trials.
+# TODO (andreyvelich): Implement list_base_images function to get each image description.
+BASE_IMAGE_TENSORFLOW = "docker.io/tensorflow/tensorflow:2.9.1"
+BASE_IMAGE_MXNET = "docker.io/mxnet/python:1.9.1_native_py3"
```
Can these be modified by the user? Otherwise, the versions here are tied to Katib Python SDK releases.

andreyvelich replied:
@terrytangyuan Sure, users can set [the custom base image](https://github.com/kubeflow/katib/blob/86d3a686f3c5d4e3e98d68465aade78d74c96b4e/sdk/python/v1beta1/kubeflow/katib/api/katib_client.py#L120).

terrytangyuan commented on README.md:
```diff
@@ -189,10 +189,51 @@ katib-ui-5767cfccdc-pwg2x           1/1     Running     0          36s
+
+# Step 2. Create HyperParameter search space.
+parameters = {
+    "a": katib.int(min=10, max=20),
+    "b": katib.double(min=0.1, max=0.2)
```
This doesn't seem Pythonic to me. Is it easy to support user specified list ranges, e.g. `range(10, 20)`?

andreyvelich replied:
@terrytangyuan Let's continue discussion around it here: https://github.com/kubeflow/katib/pull/1951#discussion_r970890433.

andreyvelich committed:
Add Final to consts
Modify packages_to_install doc
Create validate objective function

tenzen-y commented on sdk/python/v1beta1/setup.py:
```diff
@@ -14,40 +14,44 @@
+    "six>=1.10",
+    "setuptools>=21.0.0",
+    "urllib3>=1.15.1",
+    "kubernetes>=12.0.0",
```
I think that `12.0.0` is so old.
Should we change the required kubernetes client version?

https://github.com/kubernetes-client/python#compatibility-matrix-of-supported-client-versions

andreyvelich replied:
Sure, let's be in sync with Training Operator SDK.

tenzen-y commented on sdk/python/v1beta1/kubeflow/katib/constants/constants.py:
```diff
@@ -13,15 +13,28 @@
+
+# Supported base images for the Katib Trials.
+# TODO (andreyvelich): Implement list_base_images function to get each image description.
+BASE_IMAGE_TENSORFLOW: Final[str] = "docker.io/tensorflow/tensorflow:2.9.1"
```
Should we add the base image for tensorflow with GPU support same as the tfjob client?

tenzen-y reviewed (commented):
@andreyvelich Thanks for your contributions!
I left a few comments.

andreyvelich committed:
Add GPU TF Image
Change k8s version package

andreyvelich committed:
Create search module

andreyvelich commented:
Thanks everyone for review!
I addressed the remaining comments and [create separate search module for HPs](https://github.com/kubeflow/katib/pull/1951/commits/316d4a4fbf9f50f839b9141908da3198f0884f72).
Please take a look @anencore94 @tenzen-y @johnugeorge @terrytangyuan

tenzen-y commented on sdk/python/v1beta1/kubeflow/katib/api/search.py:
```diff
@@ -0,0 +1,64 @@
+# Copyright 2021 The Kubeflow Authors.
```
Probably, this is `2022`, not `2021`.

tenzen-y reviewed (commented):
@andreyvelich Thanks for updating this!

tenzen-y commented on README.md:
```diff
@@ -189,10 +189,51 @@ katib-ui-5767cfccdc-pwg2x           1/1     Running     0          36s
-- Run your first Katib Experiment in the
-  [getting started guide](https://www.kubeflow.org/docs/components/katib/hyperparameter/#example-using-random-algorithm).
+- Check
+  [the Katib getting started guide](https://www.kubeflow.org/docs/components/katib/hyperparameter/#example-using-random-algorithm).
```
```suggestion
  [the Katib getting started guide](https://www.kubeflow.org/docs/components/katib/hyperparameter/#example-using-random-search-algorithm).
```

I found this link had been broken for some time.

andreyvelich committed:
Fix link in README

Co-authored-by: Yuki Iwai <yuki.iwai.tz@gmail.com>

andreyvelich committed:
Fix licence date

tenzen-y commented:
Thanks for addressing my comments!
/lgtm

johnugeorge commented:
/approve

johnugeorge commented:
Thanks @andreyvelich for this great feature

andreyvelich commented:
Thanks everyone for your feedback and review!
I am looking forward for the new Katib usage.
/hold cancel

terrytangyuan reviewed (commented):
This is great! 

/lgtm
